docs/r2dbc.md

```diff
@@ -0,0 +1,286 @@
+
+Add the following connection properties:
+
+| Property         | Value                                                             |
+| ---------------- |-------------------------------------------------------------------|
+| DATABASE_NAME   | The name of the database to connect to                            |
+| CLOUD_SQL_CONNECTION_NAME | The instance connection name (found on the instance details page) |
+| DB_USER         | Database username                                                 |
+| DB_PASS         | Database user's password                                          |
+
+R2DBC URL template:
+
+#### Mysql
+
+```
+r2dbc:gcp:mysql://<DB_USER>:<DB_PASS>@<CLOUD_SQL_CONNECTION_NAME>/<DATABASE_NAME>
+```
+
+#### Postgres
+
+```
+r2dbc:gcp:postgres://<DB_USER>:<DB_PASS>@<CLOUD_SQL_CONNECTION_NAME>/<DATABASE_NAME>
+```
+
+##### SQL Server
+
+```
+r2dbc:gcp:mssql://<DB_USER>:<DB_PASS>@<CLOUD_SQL_CONNECTION_NAME>/<DATABASE_NAME>
+```
+
+#### Example
+
+```java
+// Set up URL parameters
+ConnectionFactoryOptions options = ConnectionFactoryOptions.builder()
+    .option(DRIVER, "gcp")
+    .option(PROTOCOL, "mysql") // OR "postgres" or "mssql"
+    .option(PASSWORD, "<DB_PASSWORD>")
+    .option(USER, "<DB_USER>")
+    .option(DATABASE, "<DATABASE_NAME}")
+    .option(HOST, "<CLOUD_SQL_CONNECTION_NAME>")
+    .build();
+
+ConnectionFactory connectionFactory = ConnectionFactories.get(options);
+ConnectionPoolConfiguration configuration = ConnectionPoolConfiguration
+    .builder(connectionFactory)
+    .build();
+
+ConnectionPool connectionPool = new ConnectionPool(configuration);
+```
+
+### IAM Authentication
+
+**Note:** This feature is currently only supported for Mysql and Postgres drivers.
+
+Connections using
+[IAM database authentication](https://cloud.google.com/sql/docs/postgres/iam-logins)
+are supported when connecting to Postgres instances.
+This feature is unsupported for other drivers. First, make sure to
+[configure your Cloud SQL Instance to allow IAM authentication](https://cloud.google.com/sql/docs/postgres/create-edit-iam-instances#configure-iam-db-instance)
+and
+[add an IAM database user](https://cloud.google.com/sql/docs/postgres/create-manage-iam-users#creating-a-database-user).
+Now, you can connect using user or service
+account credentials instead of a password.
+When setting up the connection, set the `ENABLE_IAM_AUTH` connection property to `true` and `user`
+to the email address associated with your IAM user.
+
+Note: a non-empty string value for the `password` property must be set. While this property will
+be ignored when connecting with the Cloud SQL Connector using IAM auth, leaving it empty will cause
+driver-level validations to fail.
+
+#### Mysql
+
+```java
+    // Set up ConnectionFactoryOptions
+ConnectionFactoryOptions options = ConnectionFactoryOptions.builder()
+    .option(DRIVER, "gcp")
+    .option(PROTOCOL, "mysql")
+    .option(PASSWORD, "password")
+    .option(USER, "mysql-iam-user@gmail.com")
```

Comment:
Could we make these users service account users? In MySQL, everything after the @ sign is truncated. In postgres it's just my-sa@my-proeject.iam (and no .gserviceaccount.com)

Reply:
Ditto below under impersonation.

Reply:
This is sounding more like a rewrite of the IAM and Service Account Impersonation docs. Do we want to put that in a separate PR and make r2dbc.md and jdbc.md match?